[cxx-interop] Add documentation about calling ctor or static factory of SWIFT_SHARED_REFERENCE types as Swift Initializer (#1079)
swiftlang/swift-org-website: documentation/cxx-interop/index.md

Now that `SharedObject` is imported as a reference type in Swift, the programmer will be able to use it in the following manner:
```swift
// Call the C++ constructor of SharedObject using Swift initializer syntax.
let object = SharedObject()
object.doSomething()
// `object` will be released here.
```

You can create instances of `SharedObject` directly in Swift by calling its C++ constructor through a Swift initializer.

> Alternatively, you can construct instances using a user-defined static factory function, provided that the factory function is annotated with `SWIFT_NAME("init(...)")`, where the number of `_` placeholders matches the number of parameters in the factory function

Author: > think it is worth mentioning that we use the default new/delete for allocation/deallocation and if someone wants to disable the importing of ctors as initializers they can delete these operations for their class.

Updated the patch with these changes.

Now that `SharedObject` is imported as a reference type in Swift, the programmer will be able to use it in the following manner:
```swift

> // The C++ constructor is imported as a Swift initializer

Author: Comment should be next the the line where we are calling the initializer

let object1 = SharedObject.create()
let object2 = SharedObject()
object1.doSomething()
object2.doSomething()
// `object` will be released here.
```

> You can create instances of `SharedObject` directly from Swift by calling its C++ constructor through a Swift initializer. 

Author: Create a new section here: **Contracting Shared Reference**

Reviewer: nit: should be "Constructing Shared References" (plural "References")

> You can create instances of `SharedObject` directly from Swift by calling its C++ constructor through a Swift initializer. 

Author: drop directly.

Rewrite it like this:
As illustrated in the above example, from Swift 6.2 onwards you can create instances of `SWIFT_SHARED_REFERENCE` types by calling its initializer.

> **NOTE:** Swift uses *default* `new` operator for constructing c++ shared reference types. If you'd like to prevent Swift from importing constructors as initializers, you can also delete the *default* `new` operator in C++.

Author: Note that the Swift compiler. default need not be in italics

In the event that you'd like to hide constructors you can also elete the *default* `new` operator in C++.

Author: Remove the sentence If you'd like to prevent Swift from importing constructors as initializers, you can also delete the *default* `new` operator in C++.

Reviewer: Oops I think we were looking at this at the same time. Here's my suggestion:

```suggestion
Swift uses the default `new` operator for constructing C++ shared reference types.
To prevent Swift from importing constructors as initializers, delete the `new` operator in C++.
```

> Alternatively, you can also construct instances of `SharedObject` using a user-defined static factory function, provided that the factory function is annotated with `SWIFT_NAME("init(...)")`, where the number of `_` placeholders matches the number of parameters in the factory function.

Author: You can also import a user defined static factory function as a Swift initializer by annotating it with  `SWIFT_NAME("init(...)")`, where the number of `_` placeholders matches the number of parameters in the factory function. For example ... (no need to have a new paragraph).

Reviewer: Another suggestion: replace `...` with `…` to make it clear the ellipsis should be understood as a meta-variable and not taken verbatim.

I don't think "user-defined" is necessary here.

```suggestion
You can also import a static factory function as a Swift initializer by annotating it with `SWIFT_NAME("init(…)")`, where the `init(…)` Swift signature has the same number of parameters as the C++ factory function.
For example:
```

> For example, consider a factory that performs custom allocation or returns a singleton instance:

Author: Make it only For Example,

Reviewer: ```suggestion
```

```cpp
struct SharedObject {
static SharedObject* make(int id) SWIFT_NAME("init(_:)");

  void doSomething();
} SWIFT_SHARED_REFERENCE(retainSharedObject, releaseSharedObject);
```

In this case, Swift will import the static `make` function as a Swift initializer:

```swift

> let object = SharedObject(id: 42)

Reviewer: The `id` label shouldn't be there, i.e.,

```suggestion
let object = SharedObject(42)
```

But I think you should give more examples here. Also, `doSomething()` isn't really necessary here.

Author: @j-hui Let's add more examples of using the `SWIFT_NAME` technique in a follow-up patch. We can experiment with this feature on some adopter projects or dummy examples before documenting.

object.doSomething()
```

> **Note**: If a C++ constructor and a user-annotated static factory (via `SWIFT_NAME`) have identical parameter signatures, Swift prefers the static factory when resolving initializer calls. Using `SWIFT_NAME("init(...)")` is especially useful if you want to use a custom allocator or you want to disable direct construction entirely and expose only factories. 

Author: Make this also as : Note that.

Author: Remove the note completely
